### PR TITLE
Added some generic geometry queries and associated unit tests

### PR DESCRIPTION
**Pull request by int-Frank**
In the process of creating a cylinder vs AABB query, I had to write and test a number of other queries. I pushing this work separately as this can perhaps be the start of a general and centralised geometry library. We can add to it as we need to. It might be very useful for quickly prototyping ideas, or for use in areas that aren't performance critical. 

There are a number of things to consider, but here are some decisions I've initially made:
- To include an epsilon with most comparisons. 
- To have a separate epsilon for single and double precision floating point numbers
- To avoid separate classes/structs for geometric entities (plane, lines etc...) Maybe this should change actually. Perhaps we should encapsulate objects in their own class.
- There are a few different types of queries; Distance between things, Closest points between things, Intersecting space between things. Our API should reflect this. I have opted for `udXX_YYY` where `XX` is the type of query, either `DT` (distance test), `CP` (closest points), `IT` (test for intersection), and `YYY` is a description of the query: `PointLine` for example.

### CODE REVIEW

File: Include/udGeometry.h [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+// Constants
+//------------------------------------------------------------------------
+
+#//define UD_USE_EXACT_MATH
+
+template<typename T>
+struct udEpsilon {};
+
+template<> struct udEpsilon<float> { static const float value; };
+template<> struct udEpsilon<double> { static const double value; };
```
Comment (samsinsane) [inline Include/udGeometry.h:22]: You can avoid the requirement of a cpp file by doing something similar to:
```c++
template<typename T>
constexpr T GetEpsilon();

template<>
constexpr double udGetEpsilon() { return 1e-12; }

template<>
constexpr float udGetEpsilon() { return 1e-6f; }
```

File: Include/udGeometry.h [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+enum udGeometryCode
+{
+  udGC_Fail = 0,
```
Comment (samsinsane) [inline Include/udGeometry.h:26]: This isn't necessary, this is the default behaviour.

File: Include/udGeometry.h [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  udGC_Overlapping,
+  udGC_Intersecting,
+  udGC_NotIntersecting,
+  udGC_CompletelyInside,
+  udGC_CompletelyOutside
+};
+
+//------------------------------------------------------------------------
+// Index
+//------------------------------------------------------------------------
```
Comment (samsinsane) [inline Include/udGeometry.h:38]: These comment blocks aren't necessary.

File: Include/udGeometry.h [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+template<typename T>bool udIsZero(T);
+
+template<typename T> udGeometryCode udCP_PointLine3(const udVector3<T> &lineOrigin, const udVector3<T> &lineDirection, const udVector3<T> &point, udVector3<T> &out, T *pU = nullptr);
+template<typename T> udGeometryCode udCP_PointSegment3(const udVector3<T> &ls0, const udVector3<T> &ls1, const udVector3<T> &point, udVector3<T> &out, T *pU = nullptr);
+template<typename T> udGeometryCode udCP_SegmentSegment3(const udVector3<T> &a0, const udVector3<T> &a1, const udVector3<T> &b0, const udVector3<T> &b1, udVector3<T> &aOut, udVector3<T> &bOut, udVector2<T> *pU = nullptr);
+
+//------------------------------------------------------------------------
+// Implementation
+//------------------------------------------------------------------------
```
Comment (samsinsane) [inline Include/udGeometry.h:48]: If you put all the implementations in a `_inl.h` file, similar to `udMath.h` and `udMath_inl.h`. then you don't need markers like this.

File: udTest/src/udMathTests.cpp [hunk 5]
```diff
@@ -1061,3 +1062,239 @@ TEST(MathTests, MatrixExtraction)
   EXPECT_DOUBLE_EQ(outQuatYPR.y, outMatYPR.y);
   EXPECT_DOUBLE_EQ(outQuatYPR.z, outMatYPR.z);
 }
+
+TEST(MathTests, Geometry)
```
Comment (samsinsane) [inline udTest/src/udMathTests.cpp:1066]: This should be in its own file, `udGeometryTests.cpp`.

File: Include/udGeometry.h [hunk 6]
```diff
@@ -0,0 +1,238 @@
+//
+// Copyright (c) Euclideon Pty Ltd
+//
+// Creator: Frank Hart, June 2020
+//
```
Comment (samsinsane) [inline Include/udGeometry.h:5]: I'm not sure these comment blocks are required by any of our standards.

File: Include/udGeometry_Inl.h [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+  out = ls0 + u * axis;
+
+  if (pU != nullptr)
+    *pU = u;
+
+  return udGC_Success;
+}
+
+template<typename T>
+udGeometryCode udCP_SegmentSegment3(const udVector3<T> &a0, const udVector3<T> &a1, const udVector3<T> &b0, const udVector3<T> &b1, udVector3<T> &aOut, udVector3<T> &bOut, udVector2<T> * pU)
```
Comment (dpev) [inline Include/udGeometry_Inl.h:60]: What is the udCP_ prefix?

File: Include/udGeometry.h [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+// Constants
+//------------------------------------------------------------------------
+
+//#define UD_USE_EXACT_MATH
+
+template<typename T>
+constexpr T udGetEpsilon();
+
+template<>
+constexpr double udGetEpsilon() { return 1e-12; };
```
Comment (dpev) [inline Include/udGeometry.h:16]: I like these but probably belong in udMath.h?

File: Include/udGeometry.h [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+template<>
+constexpr float udGetEpsilon() { return 1e-6f; };
+
+enum udGeometryCode
```
Comment (dpev) [inline Include/udGeometry.h:21]: This needs comments, seems like a lot of cases to have callers have to handle

File: Include/udGeometry_Inl.h [hunk 10]
```diff
@@ -0,0 +1,188 @@
+
+#ifdef UD_USE_EXACT_MATH
+template<typename T>
```
Comment (dpev) [inline Include/udGeometry_Inl.h:3]: These look better as one-liners like the epsilon functions

File: Include/udGeometry_Inl.h [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+  udVector3<T> w;
+  w = point - lineOrigin;
+  T proj = udDot(w, lineDirection);
+  out = lineOrigin + proj * lineDirection;
+  if (pU != nullptr)
+    *pU = proj;
+  return udGC_Success;
+}
+
+template<typename T>
```
Comment (dpev) [inline Include/udGeometry_Inl.h:28]: These functions should have the standard header all UD functions have too

File: Include/udGeometry_Inl.h (hunk 7 above)
Comment (int-Frank) [reply Include/udGeometry_Inl.h:60]: Closest Points. I figured we need a prefix for the different types of queries: closest points, distance to, intersection test, find intersection space. A lot of these are implemented differently

File: Include/udGeometry.h (hunk 8 above)
Comment (int-Frank) [reply Include/udGeometry.h:16]: I see there is already a UD_EPSILON in udMath.h. Will this clash?

File: Include/udGeometry_Inl.h (hunk 7 above)
Comment (dpev) [reply Include/udGeometry_Inl.h:60]: Should probably have the consistent naming scheme of udGeometry_xxx if it's a public API function. The CP is a moniker we can work on how to put on relevant functions in a way that is consistent throughout UD codebase

File: Include/udGeometry.h (hunk 8 above)
Comment (dpev) [reply Include/udGeometry.h:16]: No, but one might want to replace the other. I like the newer one becasue the epsilon is automatically correct depending on float/double

File: Include/udGeometry.h (hunk 9 above)
Comment (int-Frank) [reply Include/udGeometry.h:21]: I see... These are contextual, so individual geometry queries should also have a comment outlining what the query can return.

**Pull request closed without merging**